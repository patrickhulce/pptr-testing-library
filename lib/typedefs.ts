import {Matcher, MatcherOptions, SelectorMatcherOptions} from '@testing-library/dom'
import {ElementHandle} from 'puppeteer'

type Element = ElementHandle

interface IQueryMethods {
  queryByPlaceholderText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByPlaceholderText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByPlaceholderText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByPlaceholderText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>

  queryByText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element | null>
  queryAllByText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element[]>
  getByText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element>
  getAllByText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element[]>

  queryByLabelText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element | null>
  queryAllByLabelText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element[]>
  getByLabelText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element>
  getAllByLabelText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element[]>

  queryByAltText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByAltText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByAltText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByAltText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>

  queryByTestId(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByTestId(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByTestId(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByTestId(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>

  queryByTitle(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByTitle(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByTitle(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByTitle(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>

  queryByRole(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByRole(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByRole(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByRole(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>

  queryByDisplayValue(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByDisplayValue(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByDisplayValue(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByDisplayValue(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
}

type IScopedQueryMethods = {
  [K in keyof IQueryMethods]: (m: Matcher, opts?: MatcherOptions) => ReturnType<IQueryMethods[K]>
}

export interface IScopedQueryUtils extends IScopedQueryMethods {
  getQueriesForElement(): IQueryUtils & IScopedQueryUtils
  getNodeText(): Promise<string>
}

export interface IQueryUtils extends IQueryMethods {
  getQueriesForElement(): IQueryUtils & IScopedQueryUtils
  getNodeText(el: Element): Promise<string>
}

export interface Config {
  testIdAttribute: string;
}
