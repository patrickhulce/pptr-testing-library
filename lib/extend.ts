import {getDocument, getQueriesForElement} from '.'
import {ElementHandle} from '../node_modules/@types/puppeteer'
import {IScopedQueryUtils} from './typedefs'

// tslint:disable-next-line
let Page, ElementHandle

function requireOrUndefined(path: string): any {
  try {
    return require(path)
  } catch (err) {}
}

try {
  Page = require('puppeteer/lib/Page.js') // tslint:disable-line
  if (Page.Page) Page = Page.Page

  ElementHandle = requireOrUndefined('puppeteer/lib/ElementHandle.js') // tslint:disable-line variable-name
  if (ElementHandle && ElementHandle.ElementHandle) ElementHandle = ElementHandle.ElementHandle

  if (!ElementHandle) {
    const ExecutionContext = requireOrUndefined('puppeteer/lib/ExecutionContext.js') // tslint:disable-line variable-name
    if (ExecutionContext && ExecutionContext.ElementHandle) {
      ElementHandle = ExecutionContext.ElementHandle
    }
  }
  if (ElementHandle && ElementHandle.ElementHandle) ElementHandle = ElementHandle.ElementHandle

  if (!ElementHandle) {
    const JSHandle = require('puppeteer/lib/JSHandle.js') // tslint:disable-line
    if (JSHandle && JSHandle.ElementHandle) {
      ElementHandle = JSHandle.ElementHandle
    }
  }
  if (ElementHandle && ElementHandle.ElementHandle) ElementHandle = ElementHandle.ElementHandle

  Page.prototype.getDocument = getDocument
  getQueriesForElement(ElementHandle.prototype, function(this: ElementHandle): ElementHandle {
    return this
  })

  ElementHandle.prototype.getQueriesForElement = function(this: ElementHandle): ElementHandle {
    return getQueriesForElement(this)
  }
} catch (err) {
  // tslint:disable-next-line
  console.error('Could not augment puppeteer functions, do you have a conflicting version?')
  throw err
}

/* tslint:disable */
declare module 'puppeteer' {
  interface Page {
    getDocument(): Promise<ElementHandle>
  }

  interface ElementHandle extends IScopedQueryUtils {}
}
