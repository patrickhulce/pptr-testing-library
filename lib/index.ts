import {readFileSync} from 'fs'
import * as path from 'path'
import {ElementHandle, EvaluateFn, JSHandle, Page} from 'puppeteer'
import waitForExpect from 'wait-for-expect'
import {IQueryUtils, IScopedQueryUtils} from './typedefs'

const domLibraryAsString = readFileSync(
  path.join(__dirname, '../dom-testing-library.js'),
  'utf8',
).replace(/process.env/g, '{}')

/* istanbul ignore next */
function mapArgument(argument: any, index: number): any {
  return index === 0 && typeof argument === 'object' && argument.regex
    ? new RegExp(argument.regex, argument.flags)
    : argument
}

const delegateFnBodyToExecuteInPage = `
  ${domLibraryAsString};

  const mappedArgs = args.map(${mapArgument.toString()});
  const moduleWithFns = fnName in __dom_testing_library__ ?
    __dom_testing_library__ :
    __dom_testing_library__.__moduleExports;
  return moduleWithFns[fnName](container, ...mappedArgs);
`

type DOMReturnType = ElementHandle | ElementHandle[] | null

type ContextFn = (...args: any[]) => ElementHandle

async function createElementHandleArray(handle: JSHandle): Promise<ElementHandle[]> {
  const lengthHandle = await handle.getProperty('length')
  const length = await lengthHandle.jsonValue()

  const elements: ElementHandle[] = []
  for (let i = 0; i < length; i++) {
    const jsElement = await handle.getProperty(i.toString())
    const element = await createElementHandle(jsElement)
    if (element) elements.push(element)
  }

  return elements
}

async function createElementHandle(handle: JSHandle): Promise<ElementHandle | null> {
  const element = handle.asElement()
  if (element) return element
  await handle.dispose()
  return null // tslint:disable-line
}

async function covertToElementHandle(handle: JSHandle, asArray: boolean): Promise<DOMReturnType> {
  return asArray ? createElementHandleArray(handle) : createElementHandle(handle)
}

function processNodeText(handles: IHandleSet): Promise<string> {
  return handles.containerHandle
    .executionContext()
    .evaluate(handles.evaluateFn, handles.containerHandle, 'getNodeText')
}

async function processQuery(handles: IHandleSet): Promise<DOMReturnType> {
  const {containerHandle, evaluateFn, fnName, argsToForward} = handles

  try {
    const handle = await containerHandle
      .executionContext()
      .evaluateHandle(evaluateFn, containerHandle, fnName, ...argsToForward)
    return await covertToElementHandle(handle, fnName.includes('All'))
  } catch (err) {
    err.message = err.message.replace('[fnName]', `[${fnName}]`)
    err.stack = err.stack.replace('[fnName]', `[${fnName}]`)
    throw err
  }
}

interface IHandleSet {
  containerHandle: ElementHandle
  evaluateFn: EvaluateFn
  fnName: string
  argsToForward: any[]
}

function createDelegateFor<T = DOMReturnType>(
  fnName: keyof IQueryUtils,
  contextFn?: ContextFn,
  processHandleFn?: (handles: IHandleSet) => Promise<T>,
): (...args: any[]) => Promise<T> {
  // @ts-ignore
  processHandleFn = processHandleFn || processQuery

  const convertRegExp = (regex: RegExp) => ({regex: regex.source, flags: regex.flags})

  return async function(...args: any[]): Promise<T> {
    // @ts-ignore
    const containerHandle: ElementHandle = contextFn ? contextFn.apply(this, args) : this
    // @ts-ignore
    const evaluateFn: EvaluateFn = new Function(
      'container, fnName, ...args',
      delegateFnBodyToExecuteInPage,
    )

    // Convert RegExp to a special format since they don't serialize well
    let argsToForward = args.map(arg => (arg instanceof RegExp ? convertRegExp(arg) : arg))
    // Remove the container from the argsToForward since it's always the first argument
    if (containerHandle === args[0]) {
      argsToForward = args.slice(1)
    }

    return processHandleFn!({fnName, containerHandle, evaluateFn, argsToForward})
  }
}

export async function getDocument(_page?: Page): Promise<ElementHandle> {
  // @ts-ignore
  const page: Page = _page || this
  const documentHandle = await page.mainFrame().evaluateHandle('document')
  const document = documentHandle.asElement()
  if (!document) throw new Error('Could not find document')
  return document
}

export function wait(
  callback: () => any = () => undefined,
  {timeout = 4500, interval = 50} = {}, // tslint:disable-line
): Promise<{}> {
  return waitForExpect(callback, timeout, interval)
}

export function getQueriesForElement<T>(
  object: T,
  contextFn?: ContextFn,
): T & IQueryUtils & IScopedQueryUtils {
  const o = object as any
  if (!contextFn) contextFn = () => o
  o.getQueriesForElement = () => getQueriesForElement(o, () => o)

  o.queryByPlaceholderText = createDelegateFor('queryByPlaceholderText', contextFn)
  o.queryAllByPlaceholderText = createDelegateFor('queryAllByPlaceholderText', contextFn)
  o.getByPlaceholderText = createDelegateFor('getByPlaceholderText', contextFn)
  o.getAllByPlaceholderText = createDelegateFor('getAllByPlaceholderText', contextFn)

  o.queryByText = createDelegateFor('queryByText', contextFn)
  o.queryAllByText = createDelegateFor('queryAllByText', contextFn)
  o.getByText = createDelegateFor('getByText', contextFn)
  o.getAllByText = createDelegateFor('getAllByText', contextFn)

  o.queryByLabelText = createDelegateFor('queryByLabelText', contextFn)
  o.queryAllByLabelText = createDelegateFor('queryAllByLabelText', contextFn)
  o.getByLabelText = createDelegateFor('getByLabelText', contextFn)
  o.getAllByLabelText = createDelegateFor('getAllByLabelText', contextFn)

  o.queryByAltText = createDelegateFor('queryByAltText', contextFn)
  o.queryAllByAltText = createDelegateFor('queryAllByAltText', contextFn)
  o.getByAltText = createDelegateFor('getByAltText', contextFn)
  o.getAllByAltText = createDelegateFor('getAllByAltText', contextFn)

  o.queryByTestId = createDelegateFor('queryByTestId', contextFn)
  o.queryAllByTestId = createDelegateFor('queryAllByTestId', contextFn)
  o.getByTestId = createDelegateFor('getByTestId', contextFn)
  o.getAllByTestId = createDelegateFor('getAllByTestId', contextFn)

  o.queryByTitle = createDelegateFor('queryByTitle', contextFn)
  o.queryAllByTitle = createDelegateFor('queryAllByTitle', contextFn)
  o.getByTitle = createDelegateFor('getByTitle', contextFn)
  o.getAllByTitle = createDelegateFor('getAllByTitle', contextFn)

  o.queryByRole = createDelegateFor('queryByRole', contextFn)
  o.queryAllByRole = createDelegateFor('queryAllByRole', contextFn)
  o.getByRole = createDelegateFor('getByRole', contextFn)
  o.getAllByRole = createDelegateFor('getAllByRole', contextFn)

  o.queryByDisplayValue = createDelegateFor('queryByDisplayValue', contextFn)
  o.queryAllByDisplayValue = createDelegateFor('queryAllByDisplayValue', contextFn)
  o.getByDisplayValue = createDelegateFor('getByDisplayValue', contextFn)
  o.getAllByDisplayValue = createDelegateFor('getAllByDisplayValue', contextFn)

  o.getNodeText = createDelegateFor<string>('getNodeText', contextFn, processNodeText)

  return o
}

export const within = getQueriesForElement

// @ts-ignore
export const queries: IQueryUtils = {}
getQueriesForElement(queries, el => el)
