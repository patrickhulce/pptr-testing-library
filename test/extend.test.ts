import * as path from 'path'
import * as puppeteer from 'puppeteer'
import '../lib/extend'

describe('lib/extend.ts', () => {
  let browser: puppeteer.Browser
  let page: puppeteer.Page
  let document: puppeteer.ElementHandle

  it('should launch puppeteer', async () => {
    browser = await puppeteer.launch({args: ['--no-sandbox', '--disable-setuid-sandbox']})
    page = await browser.newPage()
    await page.goto(`file://${path.join(__dirname, 'fixtures/page.html')}`)
  })

  it('should extend puppeteer ElementHandle', async () => {
    document = await page.getDocument()
    expect(typeof document.queryAllByAltText).toBe('function')
  })

  it('should handle the query* methods', async () => {
    const element = await document.queryByText('Hello h1')
    expect(element).toBeTruthy()
    expect(await page.evaluate(el => el.textContent, element)).toEqual('Hello h1')
  })

  it('should handle regex matching', async () => {
    const element = await document.queryByText(/Hello/)
    expect(element).toBeTruthy()
    expect(await page.evaluate(el => el.textContent, element)).toEqual('Hello h1')
  })

  it('should handle the get* methods', async () => {
    const element = await document.getByTestId('testid-text-input')
    expect(await page.evaluate(el => el.outerHTML, element)).toMatchSnapshot()
  })

  it('should handle the get* method failures', async () => {
    // Use the scoped element so the pretty HTML snapshot is smaller
    const scope = await document.$('#scoped')

    try {
      await scope.getByTitle('missing')
      fail()
    } catch (err) {
      expect(err).toMatchSnapshot()
    }
  })

  it('should handle the LabelText methods', async () => {
    const element = await document.getByLabelText('Label A')
    expect(await page.evaluate(el => el.outerHTML, element)).toMatchSnapshot()
  })

  it('should handle the queryAll* methods', async () => {
    const elements = await document.queryAllByText(/Hello/)
    expect(elements).toHaveLength(3)

    const text = await Promise.all([
      page.evaluate(el => el.textContent, elements[0]),
      page.evaluate(el => el.textContent, elements[1]),
      page.evaluate(el => el.textContent, elements[2]),
    ])

    expect(text).toEqual(['Hello h1', 'Hello h2', 'Hello h3'])
  })

  it('should scope results to element', async () => {
    const scope = await document.$('#scoped')
    const element = await scope.queryByText(/Hello/)
    expect(await page.evaluate(el => el.textContent, element)).toEqual('Hello h3')
  })

  it('should get text content', async () => {
    const $h3 = await document.$('#scoped h3')
    expect(await $h3.getNodeText()).toEqual('Hello h3')
  })

  it('should work with destructuring', async () => {
    const {queryByText} = (await document.$('#scoped')).getQueriesForElement()
    expect(await queryByText('Hello h1')).toBeFalsy()
    expect(await queryByText('Hello h3')).toBeTruthy()
  })

  afterAll(async () => {
    await browser.close()
  })
})
