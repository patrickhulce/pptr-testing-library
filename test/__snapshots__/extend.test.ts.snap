// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/extend.ts should handle the LabelText methods 1`] = `"<input id=\\"label-text-input\\" type=\\"text\\">"`;

exports[`lib/extend.ts should handle the get* method failures 1`] = `
"Evaluation failed: Error: Unable to find an element with the title: missing.

[36m<div[39m
  [33mid[39m=[32m\\"scoped\\"[39m
[36m>[39m
  [0m
    [0m
  [36m<h3>[39m
    [0mHello h3[0m
  [36m</h3>[39m
  [0m
  [0m
[36m</div>[39m
    at getElementError <stack>:X:X)
    at getAllByTitle <stack>:X:X)
    at firstResultOrNull <stack>:X:X)
    at Object.getByTitle <stack>:X:X)
    at evaluateInPage <stack>:X:X)"
`;

exports[`lib/extend.ts should handle the get* methods 1`] = `"<input type=\\"text\\" data-testid=\\"testid-text-input\\">"`;
